esphome/components/shelly_dimmer/shelly_dimmer.cpp

```diff
@@ -0,0 +1,490 @@
+      continue;
+    }
+    break;
+  }
+
+  // Poll the dimmer roughly every 10s.
+  this->set_interval("poll", 10000, [this]() { this->send_command_(SHELLY_DIMMER_PROTO_CMD_POLL, nullptr, 0); });
+
+  this->send_settings_();
+  // Do an immediate poll to refresh current state.
+  this->send_command_(SHELLY_DIMMER_PROTO_CMD_POLL, nullptr, 0);
+
+  this->ready_ = true;
+}
+
+void ShellyDimmer::write_state(light::LightState *state) {
+  if (!this->ready_) {
+    return;
+  }
+
+  float brightness;
+  state->current_values_as_brightness(&brightness);
+
+  const uint16_t brightness_int = this->convert_brightness_(brightness);
+  if (brightness_int == this->brightness_) {
+    ESP_LOGV(TAG, "Not sending unchanged value");
+    return;
+  }
+  ESP_LOGD(TAG, "Brightness update: %d (raw: %f)", brightness_int, brightness);
+
+  this->send_brightness_(brightness_int);
+}
+
+bool ShellyDimmer::upgrade_firmware_() {
+  ESP_LOGW(TAG, "Starting STM32 firmware upgrade");
+  this->reset_dfu_boot_();
+
+  // Could be constexpr in c++17
+  static const auto CLOSE = [](stm32_t *stm32) { stm32_close(stm32); };
+
+  // Cleanup with RAII
+  std::unique_ptr<stm32_t, decltype(CLOSE)> stm32{stm32_init(this, STREAM_SERIAL, 1), CLOSE};
+
+  if (!stm32) {
+    ESP_LOGW(TAG, "Failed to initialize STM32");
+    return false;
+  }
+
+  // Erase STM32 flash.
+  if (stm32_erase_memory(stm32.get(), 0, STM32_MASS_ERASE) != STM32_ERR_OK) {
+    ESP_LOGW(TAG, "Failed to erase STM32 flash memory");
+    return false;
+  }
+
+  static constexpr uint32_t BUFFER_SIZE = 256;
+
+  // Copy the STM32 firmware over in 256-byte chunks. Note that the firmware is stored
+  // in flash memory so all accesses need to be 4-byte aligned.
+  uint8_t buffer[BUFFER_SIZE];
+  const uint8_t *p = STM_FIRMWARE;
+  uint32_t offset = 0;
+  uint32_t addr = stm32->dev->fl_start;
+  const uint32_t end = addr + STM_FIRMWARE_SIZE_IN_BYTES;
+
+  while (addr < end && offset < STM_FIRMWARE_SIZE_IN_BYTES) {
+    const uint32_t left_of_buffer = std::min(end - addr, BUFFER_SIZE);
+    const uint32_t len = std::min(left_of_buffer, STM_FIRMWARE_SIZE_IN_BYTES - offset);
+
+    if (len == 0) {
+      break;
+    }
+
+    std::memcpy(buffer, p, BUFFER_SIZE);
+    p += BUFFER_SIZE;
+
+    if (stm32_write_memory(stm32.get(), addr, buffer, len) != STM32_ERR_OK) {
+      ESP_LOGW(TAG, "Failed to write to STM32 flash memory");
+      return false;
+    }
+
+    addr += len;
+    offset += len;
+  }
+
+  ESP_LOGI(TAG, "STM32 firmware upgrade successful");
+
+  return true;
+}
+
+uint16_t ShellyDimmer::convert_brightness_(float brightness) {
+  // Special case for zero as only zero means turn off completely.
+  if (brightness == 0.0) {
+    return 0;
+  }
+
+  uint16_t brightness_int = static_cast<uint16_t>(brightness * (this->max_brightness_ - this->min_brightness_));
+  brightness_int = brightness_int + this->min_brightness_;
+  return std::min(brightness_int, SHELLY_DIMMER_MAX_BRIGHTNESS);
```

Comment:
`SHELLY_DIMMER_MAX_BRIGHTNESS` should be removed and the config should be validated to only allow max 1000 for `CONF_MAX_BRIGHTNESS`

```suggestion
  return remap<float, uint16_t>(brightness, 0.0f, 1.0f, this->min_brightness_, this->max_brightness_);
```

Reply:
Reference: https://esphome.io/api/helpers_8h_source.html#l00145

Reply:
```suggestion
  remap<uint16_t, float>(brightness, 0.0f, 1.0f, this->min_brightness_, this->max_brightness_);
```

Reply:
Haha woops, got those template types the wrong way around.